core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -448,6 +453,29 @@ class TimeSeriesShard(val dataset: Dataset,
 
   def indexValues(indexName: String, topK: Int): Seq[TermInfo] = partKeyIndex.indexValues(indexName, topK)
 
+  def metadata(filter: Seq[ColumnFilter], columns: Seq[String], startTime: Long, endTime: Long):
+          Set[Map[ZeroCopyUTF8String, ZeroCopyUTF8String]] = {
+    val result = new HashSet[Map[ZeroCopyUTF8String, ZeroCopyUTF8String]]
+    if (filter.isEmpty) {
+      columns.foreach(column => {
+        result += partKeyIndex.indexValues(column, 1000)
+          .map(termInfo => (ZeroCopyUTF8String(column) -> termInfo.term))
+          .toMap // For unique results - as terms can have duplicate entries
+        })
+    } else {
+      //TODO get the limit from config
+      val partIterator = partKeyIndex.partIdsFromFilters(filter, startTime, endTime, 1000)
+      while (partIterator.hasNext) {
+        val nextPartID = partIterator.next
+        val nextPart = partitions.get(nextPartID)
+        val consumer = new SeqMapMetadataConsumer(columns)
+        dataset.partKeySchema.consumeMapItems(nextPart.partKeyBase, nextPart.partKeyOffset, 0, consumer)
```

Comment:
As discussed, the issues with this approach are multiple:
1) Assumes the map is always in position 0 - which will not be true for alternative schemas!
2)Uses the SeqMapConsumer which is very memory inefficient